### PR TITLE
add support for client call timeouts

### LINKED ISSUE #14: RpcCall may leak CompletionHandler instances in call() with timeouts
the timeout arguments are used for Future.get().
however, the CompletionHandler is not removed from the underlying transport's ReplyQueue

### PR DESCRIPTION
**Pull request by radai-rosenblatt**
new flag to jrpcgen ("-timeouts"), off by default (so should be no impact on existing users).
generates methods that accept +2 arguments - long _timeoutValue, TimeUnit _timeoutUnit.
backed by a ScheduledExecutorService in ReplyQueue (simple java.util.Timers arent reliable).

also fixes #14 along the way (what was a timeout to Future.get() now uses the new mechanism)

### CODE REVIEW

**Comment by kofemann** [comment]
ok to test

File: oncrpc4j-core/src/main/java/org/dcache/xdr/RpcCall.java [hunk 1]
```diff
@@ -341,16 +352,18 @@ public void call(int procedure, XdrAble args, CompletionHandler<RpcReply, XdrTra
...
+     * @param timeoutValue timeout value. 0 means no timeout
+     * @param timeoutUnits units for timeout value
      * @return A Future representing the result of the operation.
      * @throws OncRpcException
      * @throws IOException
      * @since 2.4.0
      */
-    public <T extends XdrAble> Future<T> call(int procedure, XdrAble args, final Class<T> type)
-            throws OncRpcException, IOException {
+    public <T extends XdrAble> Future<T> call(int procedure, XdrAble args, final Class<T> type, long timeoutValue, TimeUnit timeoutUnits)
```
Comment (kofemann) [inline oncrpc4j-core/src/main/java/org/dcache/xdr/RpcCall.java]: does this method really makes sense? The Future already has a get with timeout. I think

call(int procedure, XdrAble args, CompletionHandler<RpcReply, XdrTransport> callback, long timeoutValue, TimeUnit timeoutUnits)

is sufficient. Especially it CompletionHandler#fail will be called with TimeoutException.

Do you have some example in mind?

Comment (radai-rosenblatt) [inline oncrpc4j-core/src/main/java/org/dcache/xdr/RpcCall.java]: do you mean unregistering the callback from the completionHandler's failure method? if so we would need to wrap client-provided handlers with our own handler to do timeouts.

Comment (kofemann) [inline oncrpc4j-core/src/main/java/org/dcache/xdr/RpcCall.java]: I think there are two use cases:  code actively calls Future#get, or you register a ballback. In the second case, you want to be notified only with in a time window. I don't see a reason to have a Future and timeout at the same time.

Comment (radai-rosenblatt) [inline oncrpc4j-core/src/main/java/org/dcache/xdr/RpcCall.java]: ok, but if you call future(timeout) and get a timeout, your underlying callback handler should still be unregistered.

i'll fix and update

Comment (kofemann) [inline oncrpc4j-core/src/main/java/org/dcache/xdr/RpcCall.java]: Sure, we have to unregister callback on timeout.

Thanks!

Comment (radai-rosenblatt) [inline oncrpc4j-core/src/main/java/org/dcache/xdr/RpcCall.java]: i've updated the code (havent fixed gen and tests yet). is this version better?

Comment (kofemann) [inline oncrpc4j-core/src/main/java/org/dcache/xdr/RpcCall.java]: Yes, looks good to me (it's a bit weird to expose xid though).

Comment (radai-rosenblatt) [inline oncrpc4j-core/src/main/java/org/dcache/xdr/RpcCall.java]: the delegating Future needs the xid to unregister the callback ....
i could make the current call something like private int callInternal() and make the public call() void again ...

Comment (kofemann) [inline oncrpc4j-core/src/main/java/org/dcache/xdr/RpcCall.java]: thanks for your effort!

Comment (radai-rosenblatt) [inline oncrpc4j-core/src/main/java/org/dcache/xdr/RpcCall.java]: done

**Comment by kofemann** [comment]
do you want to squash last 5 commits into a single one?

**Comment by radai-rosenblatt** [comment]
sure. just give me 10 minutes to learn how to do it :-)

**Comment by radai-rosenblatt** [comment]
ok, squashed into the original commit

**Comment by kofemann** [comment]
thanks!

**Pull request merged**